File: code-tiny/tools/project_topology/parsers/ini.py
from __future__ import annotations

from typing import Dict, List

from ..models import (
    ConfidenceLevel,
    DescriptorFact,
    DescriptorParseOutput,
    DescriptorRole,
    DescriptorType,
    ParseDepth,
    SourceEvidence,
    safe_summary,
)
from .common import module_path_for_file


def parse_ini(*, project_id: str, path: str, text: str) -> DescriptorParseOutput:
    entries: List[Dict[str, object]] = []
    for line_number, raw_line in enumerate(text.splitlines(), 1):
        line = raw_line.strip()
        if not line or line.startswith(("#", ";")) or ":" not in line:
            continue
        key, value = line.split(":", 1)
        entries.append({"key": key.strip(), "value": value.strip(), "line": line_number})
    descriptor = DescriptorFact.create(
        project_id=project_id,
        module_path=module_path_for_file(path),
        path=path,
        descriptor_type=DescriptorType.RUNTIME_CONFIG,
        role=DescriptorRole.CONFIGURATION,
        parser="ini",
        parse_depth=ParseDepth.IDENTITY,
        summary=safe_summary(f"Flat INI configuration with {len(entries)} entries"),
        properties={"entries": entries, "format": "key-value-colon"},
        confidence=ConfidenceLevel.HIGH,
        evidence=(SourceEvidence(path, 1),),
    )
    return DescriptorParseOutput(descriptor=descriptor, dependencies=(), diagnostics=())

File: rust/crates/analyzer-topology/src/parsers/ini.rs
//! Port `parsers/ini.py` — flat key:value INI descriptor.

use crate::models::{
    confidence, descriptor_role, descriptor_type, parse_depth, safe_summary, DescriptorFact,
    PyValue, SourceEvidence,
};
use crate::parsers::common::module_path_for_file;
use crate::parsers::DescriptorParseOutput;

pub fn parse_ini(project_id: &str, path: &str, text: &str) -> DescriptorParseOutput {
    let mut entries: Vec<PyValue> = Vec::new();
    for (line_number_value, raw_line) in text.split('\n').enumerate() {
        let raw = raw_line.strip_suffix('\r').unwrap_or(raw_line);
        let line = raw.trim();
        if line.is_empty()
            || line.starts_with('#')
            || line.starts_with(';')
            || !line.contains(':')
        {
            continue;
        }
        let Some((key, value)) = line.split_once(':') else {
            continue;
        };
        let mut entry = PyValue::dict();
        entry.set("key", PyValue::Str(key.trim().to_string()));
        entry.set("value", PyValue::Str(value.trim().to_string()));
        entry.set("line", PyValue::Int(line_number_value as i64 + 1));
        entries.push(entry);
    }
    let mut properties = PyValue::dict();
    properties.set("entries", PyValue::List(entries.clone()));
    properties.set("format", PyValue::Str("key-value-colon".to_string()));
    let summary = safe_summary(&format!(
        "Flat INI configuration with {} entries",
        entries.len()
    ));
    let descriptor = DescriptorFact::create(
        project_id,
        &module_path_for_file(path),
        path,
        descriptor_type::RUNTIME_CONFIG,
        descriptor_role::CONFIGURATION,
        "ini",
        parse_depth::IDENTITY,
        summary,
        properties,
        confidence::HIGH,
        vec![SourceEvidence::at_line(path, 1)],
        vec![],
    )
    .expect("ini descriptor");
    DescriptorParseOutput {
        descriptor,
        dependencies: vec![],
        endpoints: vec![],
        diagnostics: vec![],
    }
}
